cpp/lib/connection/TypeDBDriver.cpp

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright (C) 2022 Vaticle
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include "typedb/connection/TypeDBDriver.hpp"
+
+using namespace TypeDB;
+
+namespace TypeDB {
+
+TypeDBDriver::TypeDBDriver(const std::string& coreAddress) : TypeDBDriver(TypeDBNative::connection_open_core(coreAddress.c_str())) { }
+
+TypeDBDriver::TypeDBDriver(TypeDBNative::Connection* conn) noexcept
+    : connectionNative(conn),
+      databaseManager(this->connectionNative),
+      userManager(this->connectionNative) {}
+
+TypeDBDriver::~TypeDBDriver() {
+    // TODO: connection_close is called before the databaseManager & UserManager are destructed. Should we wrap it in an object to enforce the order?
+    TypeDBNative::connection_close(connectionNative);
```

Comment:
Do you think this (the TODO) will be a problem? Closing the native connection and then the databaseManager & userManager?
Do you think it's worth introducing a really thin struct (possibly through a macro) just for this?

Reply:
It won't be a problem, there's no lifetime dependency between those three.